src/rng.rs

```diff
@@ -0,0 +1,175 @@
+    fn gen(&mut self) -> Result<W, ErrorKind>;
+    fn gen_range(&mut self, low: W, high: W) -> Result<W, ErrorKind>;
+    fn fill(&mut self, dest: &mut [W]) -> Result<(), ErrorKind>;
+}
+
+pub struct Rng {
+    rb: RNG,
+}
+
+impl Rng {
+    /// Returns 32 bits of randomness, or error
+    pub fn next(&mut self) -> Result<u32, ErrorKind> {
+        loop {
+            let status = self.rb.sr.read();
+            if status.cecs().bit() {
+                return Err(ErrorKind::ClockError);
+            }
+            if status.secs().bit() {
+                return Err(ErrorKind::SeedError);
+            }
+            if status.drdy().bit() {
+                return Ok(self.rb.dr.read().rndata().bits());
+            }
+        }
+    }
+
+    pub fn release(self) -> RNG {
+        self.rb
+    }
+
+    /// Return random boolean, or error
+    pub fn gen_bool(&mut self) -> Result<bool, ErrorKind> {
+        let val = self.next()?;
+        Ok(val & 1 == 1)
+    }
+
+    /// Return random booleans in the ratio one:zero = n:(d-n), or
+    /// error
+    pub fn gen_ratio(
+        &mut self,
+        numerator: u32,
+        denominator: u32,
+    ) -> Result<bool, ErrorKind> {
+        assert!(denominator > 0);
+        let val = self.gen_range(0, denominator)?;
+        Ok(numerator > val)
+    }
+
+    /// Choose random value from array, or error
+    pub fn choose<'a, T>(
+        &mut self,
+        values: &'a [T],
+    ) -> Result<&'a T, ErrorKind> {
+        let val = self.gen_range(0, values.len())?;
+        Ok(&values[val])
+    }
+
+    /// Choose random value from array, or error
+    pub fn choose_mut<'a, T>(
+        &mut self,
+        values: &'a mut [T],
+    ) -> Result<&'a mut T, ErrorKind> {
+        let val = self.gen_range(0, values.len())?;
+        Ok(&mut values[val])
+    }
+
+    /// Choose random value from array, or return error
+    pub fn shuffle<T>(&mut self, values: &mut [T]) -> Result<(), ErrorKind> {
+        for i in (1..values.len()).rev() {
+            values.swap(i, self.gen_range(0, i + 1)?);
+        }
+        Ok(())
+    }
+}
+
+impl rng::Read for Rng {
+    type Error = ErrorKind;
+
+    fn read(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        self.fill(buffer)
+    }
+}
+
+macro_rules! rng_core {
+    ($($type:ty),+) => {
+        $(
+            impl RngCore<$type> for Rng {
+                /// Returns a single element with random value, or error
+                fn gen(&mut self) -> Result<$type, ErrorKind> {
+                    let val = self.next()?;
+                    Ok(val as $type)
+                }
+
+                /// Returns a single element with random value in range, or error
+                fn gen_range(&mut self, low: $type, high: $type) -> Result<$type, ErrorKind> {
+                    assert!(high > low);
+                    let range = high - low;
+                    let val: $type = self.gen()?;
+                    Ok(low + val % range)
```

Comment:
Are you sure this is correct i.e. unbiased/"random"? This `gen_range(0u8, 196u8)` will return a value from 0-63 more often than 64-195.
Random numbers are not easy. I would gravitate towards passing any functionality that looks remotely complicated to a special random crate that knows how to do that.

Reply:
@jordens could you explain why `gen_range(0u8, 196u8)` will return a value from 0-63 more often than 64-195?

Reply:
From the possible values of a `u8`, both ranges `0:64` and `196:256` will be mapped to `0:64`, while only `64:196` will be mapped to `64:196`. Each number in `0:64` will have `1/128` probability of occurring, while each number in `64:196` has only `1/256` probability of occurring.

Reply:
It'd certainly be preferrable for these functions to be implemented on a trait in `embedded-hal`, or in a specialised crate. In the meantime they could go here and in `g0xx-hal` as a demonstration?

Worth fixing the modulo bias either way.

Reply:
I would rather drop that higher level functionality. It's currently wrong. It also seems unwise to commit to providing functionality in the HAL that has not been abstracted or may never be abstracted. It defeats the purpose of the HAL and chains this crate to providing that functionality while `embedded-hal` may move on. You can't really remove it without breaking people's code and it's rather likely that the API will be changed when `embedded-hal`-ified. I would avoid collecting unproven functionality in the HAL that's outside `embedded-hal`.

Reply:
Removed in another commit

Reply:
Thanks!